ui/StatusQ/src/fastexpressionsorter.cpp

```diff
@@ -50,64 +50,81 @@ void FastExpressionSorter::setExpression(const QQmlScriptString& scriptString)
     updateExpression();
 
     emit expressionChanged();
-    invalidate();
+    queueInvalidate();
+}
+
+void FastExpressionSorter::queueInvalidate()
+{
+    if (m_queuedInvalidate)
+        return;
+
+    m_queuedInvalidate = true;
+    QMetaObject::invokeMethod(this, &FastExpressionSorter::invalidate, Qt::QueuedConnection);
+}
+
+void FastExpressionSorter::onInvalidate()
+{
+    m_queuedInvalidate = false;
+    Sorter::invalidate();
 }
 
 void FastExpressionSorter::proxyModelCompleted(const QQmlSortFilterProxyModel& proxyModel)
 {
     updateContext(proxyModel);
+    Sorter::proxyModelCompleted(proxyModel);
 }
 
 /*!
     \qmlproperty list<string> FastExpressionSorter::expectedRoles
 
     List of role names intended to be available in the expression's context.
 */
-void FastExpressionSorter::setExpectedRoles(const QStringList& expectedRoles)
+void FastExpressionSorter::setExpectedRoles(const QSet<QByteArray>& expectedRoles)
 {
     if (m_expectedRoles == expectedRoles)
         return;
 
     m_expectedRoles = expectedRoles;
     emit expectedRolesChanged();
 
-    invalidate();
+    queueInvalidate();
 }
 
-const QStringList &FastExpressionSorter::expectedRoles() const
+const QSet<QByteArray> &FastExpressionSorter::expectedRoles() const
 {
     return m_expectedRoles;
 }
 
-bool evaluateBoolExpression(QQmlExpression& expression)
+int evaluateIntExpression(QQmlExpression& expression)
 {
     QVariant variantResult = expression.evaluate();
     if (expression.hasError()) {
         qWarning() << expression.error();
-        return false;
+        return -1;
     }
 
-    if (variantResult.canConvert<bool>())
-        return variantResult.toBool();
+    if (variantResult.canConvert<int>())
+        return variantResult.toInt();
 
     qWarning("%s:%i:%i : Can't convert result to bool",
              expression.sourceFile().toUtf8().data(),
              expression.lineNumber(),
              expression.columnNumber());
-    return false;
+    return -1;
```

Comment:
shouldn't it be `0`? With `-1` it's inconsistent when swapping arguments.

Reply:
Good point! I think it should be 0